src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java

```diff
@@ -290,21 +387,96 @@ protected int getMaxAttacksUntilRestock()
+            {
+                return getNextPatrollingTarget((BuildingGuardTower) building);
+            }
         }
 
-        if (worker.isWorkerAtSiteWithMove(currentPatrolTarget, PATH_CLOSE))
+        return AIState.GUARD_SEARCH_TARGET;
+    }
+
+    /**
+     * Retrieves the next patrolling target from the guard tower.
+     * @param building his building.
+     * @return the next state to go to.
+     */
+    private AIState getNextPatrollingTarget(final BuildingGuardTower building)
+    {
+        if (building.shallPatrolManually() && building.getTask().equals(BuildingGuardTower.Task.PATROL))
         {
-            currentPatrolTarget = null;
+            final BlockPos pos = building.getNextPatrolTarget(currentPathTarget);
+            if (pos != null)
+            {
+                currentPathTarget = pos;
+                return AIState.GUARD_SEARCH_TARGET;
+            }
         }
-
+        else if(building.getTask().equals(BuildingGuardTower.Task.GUARD))
+        {
+            BlockPos pos = building.getGuardPos();
+            if(pos == null)
+            {
+                pos = building.getLocation();
+            }
+            currentPathTarget = pos;
+            return AIState.GUARD_SEARCH_TARGET;
+        }
+        else if(building.getTask().equals(BuildingGuardTower.Task.FOLLOW))
+        {
+            BlockPos pos = building.getPlayerToFollow();
+            if(pos == null || BlockPosUtil.getDistance2D(pos, building.getColony().getCenter()) > Configurations.workingRangeTownHall + Configurations.townHallPadding)
+            {
+                if(pos != null)
+                {
+                    Log.getLogger().info(BlockPosUtil.getDistance2D(pos, building.getColony().getCenter()));
+                }
+                final EntityPlayer player = building.getPlayer();
+                if(player != null)
+                {
+                    LanguageHandler.sendPlayerMessage(building.getPlayer(), LanguageHandler.format("com.minecolonies.job.guard.switch"));
+                }
+                pos = building.getLocation();
+                building.setTask(BuildingGuardTower.Task.GUARD);
+            }
+            currentPathTarget = pos;
+            return AIState.GUARD_SEARCH_TARGET;
+        }
+        currentPathTarget = getRandomBuilding();
         return AIState.GUARD_SEARCH_TARGET;
     }
 
     /**
+     * Checks if the the worker is too far from his patrol/guard/follow target.
+     * @param target an attack target.
+     * @param range the range allowed to be from the patrol/guard/follow target.
+     * @return true if too far.
+     */
+    public boolean shouldReturnToTarget(final BlockPos target, final double range)
+    {
+        final AbstractBuilding building = getOwnBuilding();
+        if (building != null && building instanceof BuildingGuardTower)
+        {
+            if (BlockPosUtil.getDistance2D(target, currentPathTarget) > range)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Merge this if statement with the enclosing one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1066)